### PR TITLE
[metrics] Fix lost connection when metrics query database

### PR DESCRIPTION
**Pull request by wing2fly**
## What changes were proposed in this pull request?

Lost connection happens in Django when using multiprocessing or other parallel execution methods. It often occurs because MySQL connections are not shared safely between processes or threads, leading to disconnections or dropped connections during execution. Metrics use multiprocessing when we started Gunicorn server.

## How was this patch tested?

Manually tested on DC cluster

Please review [Hue Contributing Guide](https://github.com/cloudera/hue/blob/master/CONTRIBUTING.md) before opening a pull request.

### CODE REVIEW

**Review by quadoss — APPROVED** [review]
@wing2fly , The changes look good but there are test failures, please check.

File: apps/useradmin/src/useradmin/metrics.py [hunk 1]
```diff
@@ -32,6 +34,14 @@ def active_users():
         first_login=False,
         hostname__isnull=False
     ).count()
+  except OperationalError as oe:
+    connection.close()
+    connection.connect()
```
Comment (Harshg999) [inline apps/useradmin/src/useradmin/metrics.py:39]: Curious: Shouldn't we close the connection again after collecting the below metric?

File: apps/useradmin/src/useradmin/metrics.py [hunk 2]
```diff
@@ -51,7 +61,13 @@ def active_users():
 def active_users_per_instance():
   from useradmin.models import UserProfile
   try:
-    count = UserProfile.objects.filter(last_activity__gt=datetime.now() - timedelta(hours=1), hostname=get_localhost_name()).count()
+    count = UserProfile.objects.filter(last_activity__gt=datetime.now() - timedelta(hours=1),
+                                       hostname=get_localhost_name()).count()
+  except OperationalError as oe:
+    connection.close()
+    connection.connect()
+    count = UserProfile.objects.filter(last_activity__gt=datetime.now() - timedelta(hours=1),
```
Comment (Harshg999) [inline apps/useradmin/src/useradmin/metrics.py:69]: Same here?

File: desktop/core/src/desktop/metrics.py [hunk 3]
```diff
@@ -145,6 +147,10 @@ def user_count():
   users = 0
   try:
     users = User.objects.count()
+  except OperationalError as oe:
+    connection.close()
+    connection.connect()
```
Comment (Harshg999) [inline desktop/core/src/desktop/metrics.py:152]: Same?

File: desktop/core/src/desktop/metrics.py [hunk 4]
```diff
@@ -196,6 +202,14 @@ def num_of_queries():
       is_history=True,
       last_modified__gt=datetime.now() - timedelta(minutes=10)
     ).count()
+  except OperationalError as oe:
+    connection.close()
+    connection.connect()
```
Comment (Harshg999) [inline desktop/core/src/desktop/metrics.py:207]: Same?

**Review by Harshg999 — COMMENTED** [review]
Thanks for taking this up!

File: apps/useradmin/src/useradmin/metrics.py (hunk 1 above)
Comment (wing2fly) [reply apps/useradmin/src/useradmin/metrics.py:39]: no, if the we have other db calls

File: desktop/core/src/desktop/metrics.py (hunk 3 above)
Comment (wing2fly) [reply desktop/core/src/desktop/metrics.py:152]: no, if the we have other db calls

File: apps/useradmin/src/useradmin/metrics.py (hunk 2 above)
Comment (wing2fly) [reply apps/useradmin/src/useradmin/metrics.py:74]: no, if the we have other db calls

**Review by athithyaaselvam — APPROVED** [review]
Looks good!

File: apps/useradmin/src/useradmin/metrics.py (hunk 1 above)
Comment (amitsrivastava) [reply apps/useradmin/src/useradmin/metrics.py:41]: Can `connection.close()` itself throw an exception?

**Pull request merged**